### ISSUE TITLE
Different behavior in map command for imported vim9 functions

### ISSUE THREAD

**Issue by dvogel** [vim/vim#10656]
### Steps to reproduce

In `demo9.vim` place:
```
vim9script
export def ReturnA(): string
    return "A"
enddef
defcompile
```

In `main.vim` place:
```
set nocompatible
import "./demo9.vim" as demo9
function s:ReturnB()
    return "B"
endfunction
echo s:demo9.ReturnA()
let s:ReturnA = s:demo9.ReturnA
nmap <F3> :echo <SID>ReturnB()<CR>
nmap <F4> :echo <SID>ReturnA()<CR>
nmap <F5> :echo <SID>demo9.ReturnA()<CR>
```

Run vim as:
```
vim --clean -u main.vim
```

On startup it will echo A
F3 will echo B
F4 will echo A
F5 will yield E121 Undefined variable <SNR>1_demo9

### Expected behaviour

F3 should echo B
F4 should echo A
F5 should echo A

The `vim9.txt` help says:

```
`:import` can also be used in legacy Vim script.  The imported items still
become script-local, even when the "s:" prefix is not given.
```

This seems to say that the import alias will create a local variable `s:demo9`. That seems to be verified by the call:

```
echo s:demo9.ReturnA()
```

The ability reference script locals in a `map` command seems to be verified by:

```
nmap <F3> :echo <SID>ReturnB()<CR>
```

The ability to call compiled functions via script local references seems to be verified by:

```
let s:ReturnA = s:demo9.ReturnA
nmap <F4> :echo <SID>ReturnA()<CR>
```

The F5 map command combines all of these but has a different result.

### Version of Vim

9.0.0030

### Environment

debian/sid, TERM=xterm-kitty

VIM - Vi IMproved 9.0 (2022 Jun 28, compiled Jul  2 2022 16:32:10)
Included patches: 1-30
Compiled by dvogel@bitbox
Huge version with GTK3 GUI.  Features included (+) or not (-):
+acl               +file_in_path      +mouse_urxvt       -tag_any_white
+arabic            +find_in_path      +mouse_xterm       -tcl
+autocmd           +float             +multi_byte        +termguicolors
+autochdir         +folding           +multi_lang        +terminal
-autoservername    -footer            -mzscheme          +terminfo
+balloon_eval      +fork()            +netbeans_intg     +termresponse
+balloon_eval_term +gettext           +num64             +textobjects
+browse            -hangul_input      +packages          +textprop
++builtin_terms    +iconv             +path_extra        +timers
+byte_offset       +insert_expand     -perl              +title
+channel           +ipv6              +persistent_undo   +toolbar
+cindent           +job               +popupwin          +user_commands
+clientserver      +jumplist          +postscript        +vartabs
+clipboard         +keymap            +printer           +vertsplit
+cmdline_compl     +lambda            +profile           +vim9script
+cmdline_hist      +langmap           -python            +viminfo
+cmdline_info      +libcall           +python3           +virtualedit
+comments          +linebreak         +quickfix          +visual
+conceal           +lispindent        +reltime           +visualextra
+cryptv            +listcmds          +rightleft         +vreplace
+cscope            +localmap          +ruby              +wildignore
+cursorbind        -lua               +scrollbind        +wildmenu
+cursorshape       +menu              +signs             +windows
+dialog_con_gui    +mksession         +smartindent       +writebackup
+diff              +modify_fname      +sodium            +X11
+digraphs          +mouse             -sound             -xfontset
+dnd               +mouseshape        +spell             +xim
-ebcdic            +mouse_dec         +startuptime       +xpm
+emacs_tags        -mouse_gpm         +statusline        +xsmp_interact
+eval              -mouse_jsbterm     -sun_workshop      +xterm_clipboard
+ex_extra          +mouse_netterm     +syntax            -xterm_save
+extra_search      +mouse_sgr         +tag_binary        
-farsi             -mouse_sysmouse    -tag_old_static

**Comment by lacygoill** [comment]
As a (temporary?) workaround, you can use [`<ScriptCmd>`](https://vimhelp.org/map.txt.html#25ScriptCmd25):

    nmap <F3> <ScriptCmd>echo s:ReturnB()<CR>
    nmap <F4> <ScriptCmd>echo s:ReturnA()<CR>
    nmap <F5> <ScriptCmd>echo s:demo9.ReturnA()<CR>

As a benefit, it lets you drop `<SID>` (which is specific to a mapping) in favor of `s:`, which is what you would write at the script-level.

**Comment by brammool** [comment]
> ### Steps to reproduce
> 
> In `demo9.vim` place:
> ```
> vim9script
> export def ReturnA(): string
>     return "A"
> enddef
> defcompile
> ```
> 
> In `main.vim` place:
> ```
> set nocompatible
> import "./demo9.vim" as demo9
> function s:ReturnB()
>     return "B"
> endfunction
> echo s:demo9.ReturnA()
> let s:ReturnA = s:demo9.ReturnA
> nmap <F3> :echo <SID>ReturnB()<CR>
> nmap <F4> :echo <SID>ReturnA()<CR>
> nmap <F5> :echo <SID>demo9.ReturnA()<CR>
> ```
> 
> Run vim as:
> ```
> vim --clean -u main.vim
> ```
> 
> On startup it will echo A
> F3 will echo B
> F4 will echo A
> F5 will yield E121 Undefined variable <SNR>1_demo9
> 
> 
> ### Expected behaviour
> 
> F3 should echo B
> F4 should echo A
> F5 should echo A
> 
> The `vim9.txt` help says:
> 
> ```
> `:import` can also be used in legacy Vim script.  The imported items still
> become script-local, even when the "s:" prefix is not given.
> ```
> 
> This seems to say that the import alias will create a local variable `s:demo9`. That seems to be verified by the call:
> 
> ```
> echo s:demo9.ReturnA()
> ```

It's not actually a variable but a namespace, but it sort-of behaves
like a local read-only dictionary variable.

> The ability reference script locals in a `map` command seems to be verified by:
> 
> ```
> nmap <F3> :echo <SID>ReturnB()<CR>
> ```

This only works for a function, not for a script variable.  For example:

	let s:varC = 'C'
	nmap <F6> :echo <SID>varC<CR>

Also cannot access "varC".

> The ability to call compiled functions via script local references
> seems to be verified by:
> 
> ```
> let s:ReturnA = s:demo9.ReturnA
> nmap <F4> :echo <SID>ReturnA()<CR>
> ```
> 
> The F5 map command combines all of these but has a different result.

The best solution is to use <ScriptCmd>, this sets the context of the
command to where the mapping is defined.  This works:
	nmap <F8> <ScriptCmd>echo s:demo9.ReturnA()<CR>

I don't think we want to make script-local variables visible to a
mapping.  If needed you can use a function to access them.

-- 
Well, you come from nothing, you go back to nothing...  What have you
lost?  Nothing!
				-- Monty Python: The life of Brian

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by dvogel** [comment]
Thank you for the clarification. The term "namespace" made my think `demo9` was a part of the `ReturnA` function's fully-qualified name and thus `demo9.ReturnA` was a single name referencing a function. I thought I had ruled out the possibility of a discrete lookup for `demo9` prior to the `.` operator taking effect because I tried:

```
echo s:demo9
```

That resulted in:

```
E1060: Expected dot after name: s:demo9
```

I'll see if I can reword some of vim9.txt to clarify this.
